File: rust/crates/wylde-harness/src/model_registry/service_manifests.rs
//! Read `models` declarations from service manifests. Rust port of
//! `Core/harness/model_registry/_service_manifests.py`.
//!
//! Each service in the out-of-tree `Services/` bucket may ship a
//! `manifest.json` with a `models` array. This module collects every
//! such declaration and returns:
//!
//! * an `overrides` map (model id → kind) — wins over the heuristic.
//! * a `required_by` map (model id → list of service names) — surfaced
//!   on the entry for the GUI.
//!
//! Manifests without a `models` key are silently skipped.
//!
//! ## Scan roots are *discovered*, not listed (#125)
//!
//! The scan used to walk a hand-kept list of pre-cutover top-level folder
//! names (`Voice`, `Gateway`, `N8N`, …), most of which no longer exist,
//! and it did **not** include `Services/` — so a `Services/<svc>/manifest.json`
//! that declared a model was invisible to the model registry, silently, by
//! construction (a manifest with no `models` key is legitimately skipped, so an
//! absent root looked identical to a no-op). The roots now come from
//! [`wylde_stack::roster::discovered_folders`] — the *same* `Services/` walk
//! the updater, launcher, and lifecycle daemon already follow — so a service
//! dropped into the bucket is covered with no edit here, and honours the same
//! `WYLDE_SERVICES` override.

use std::collections::HashMap;
use std::path::{Path, PathBuf};
use std::sync::Mutex;
use std::time::SystemTime;

use once_cell::sync::Lazy;
use serde_json::Value;

use crate::model_registry::types::Kind;

type Signature = Vec<(PathBuf, u64, u64)>;

struct Cache {
    signature: Option<Signature>,
    overrides: HashMap<String, Kind>,
    required_by: HashMap<String, Vec<String>>,
}

static CACHE: Lazy<Mutex<Cache>> = Lazy::new(|| {
    Mutex::new(Cache {
        signature: None,
        overrides: HashMap::new(),
        required_by: HashMap::new(),
    })
});

/// Wylde repo root. Mirrors Python's `parents[3]` walk from
/// `model_registry/_service_manifests.py`. Since we can't derive a
/// similar path from a Rust source file at runtime, fall back to
/// `WYLDE_ROOT` env var (set by Lifecycle when it spawns us) or `cwd`.
pub fn wylde_root() -> PathBuf {
    std::env::var_os("WYLDE_ROOT")
        .map(PathBuf::from)
        .unwrap_or_else(|| PathBuf::from("."))
}

/// Find every service `manifest.json` in the `Services/` bucket.
///
/// The roots are *discovered*, not listed:
/// [`wylde_stack::roster::discovered_folders`] returns each
/// immediate `Services/<svc>/` folder that carries a readable `manifest.json`
/// (dot/underscore-prefixed folders excluded), the same walk the updater,
/// launcher, and lifecycle registry use. Each such folder's `manifest.json` is
/// a candidate; whether it declares any `models` is decided later in
/// [`read_one`].
fn candidate_manifests() -> Vec<PathBuf> {
    let root = wylde_root();
    let mut paths: Vec<PathBuf> = Vec::new();
    let mut seen: std::collections::HashSet<PathBuf> = std::collections::HashSet::new();
    for folder in wylde_stack::roster::discovered_folders(&root) {
        let manifest = folder.join("manifest.json");
        // `discovered_folders` already guarantees the manifest exists, but the
        // re-check keeps this robust to a folder that vanishes between the walk
        // and here (and is cheap).
        if manifest.is_file() && seen.insert(manifest.clone()) {
            paths.push(manifest);
        }
    }
    paths
}

fn mtime_epoch(meta: &std::fs::Metadata) -> u64 {
    meta.modified()
        .ok()
        .and_then(|t| t.duration_since(SystemTime::UNIX_EPOCH).ok())
        .map(|d| d.as_secs())
        .unwrap_or(0)
}

fn scan_signature(paths: &[PathBuf]) -> Signature {
    let mut sigs: Signature = Vec::with_capacity(paths.len());
    for p in paths {
        let Ok(meta) = std::fs::metadata(p) else {
            continue;
        };
        sigs.push((p.clone(), mtime_epoch(&meta), meta.len()));
    }
    sigs.sort();
    sigs
}

fn coerce_kind(value: &Value) -> Option<Kind> {
    value.as_str().and_then(Kind::parse)
}

fn read_one(manifest: &Path) -> (String, Vec<(String, Kind)>) {
    let parent_name = manifest
        .parent()
        .and_then(|p| p.file_name())
        .and_then(|n| n.to_str())
        .unwrap_or("")
        .to_owned();
    let Ok(raw) = std::fs::read_to_string(manifest) else {
        return (parent_name, Vec::new());
    };
    let Ok(data) = serde_json::from_str::<Value>(&raw) else {
        return (parent_name, Vec::new());
    };
    let Some(map) = data.as_object() else {
        return (parent_name, Vec::new());
    };
    let service = map
        .get("name")
        .and_then(Value::as_str)
        .map(str::to_owned)
        .unwrap_or(parent_name);
    let Some(models_raw) = map.get("models").and_then(Value::as_array) else {
        return (service, Vec::new());
    };
    let mut out: Vec<(String, Kind)> = Vec::new();
    for spec in models_raw {
        let Some(spec_map) = spec.as_object() else {
            continue;
        };
        let model_id = spec_map
            .get("id")
            .and_then(Value::as_str)
            .map(str::trim)
            .unwrap_or("")
            .to_owned();
        let Some(kind) = spec_map.get("kind").and_then(coerce_kind) else {
            continue;
        };
        if model_id.is_empty() {
            continue;
        }
        out.push((model_id, kind));
    }
    (service, out)
}

fn build_declarations() -> (HashMap<String, Kind>, HashMap<String, Vec<String>>) {
    let mut overrides: HashMap<String, Kind> = HashMap::new();
    let mut required_by: HashMap<String, Vec<String>> = HashMap::new();
    for manifest in candidate_manifests() {
        let (service, decls) = read_one(&manifest);
        for (model_id, kind) in decls {
            match overrides.get(&model_id) {
                Some(existing) if *existing != kind => {
                    // First service to declare wins, matching Python's
                    // "keeping first" semantics.
                }
                Some(_) | None => {
                    overrides.entry(model_id.clone()).or_insert(kind);
                }
            }
            required_by
                .entry(model_id)
                .or_default()
                .push(service.clone());
        }
    }
    (overrides, required_by)
}

/// Return cached `(overrides, required_by)` from all service manifests.
/// Cached on the manifests' `(path, mtime, size)` signature, same
/// template as the HF scanner. `force=true` rebuilds.
pub fn load_declarations(force: bool) -> (HashMap<String, Kind>, HashMap<String, Vec<String>>) {
    let paths = candidate_manifests();
    let sig = scan_signature(&paths);
    let mut guard = match CACHE.lock() {
        Ok(g) => g,
        Err(p) => p.into_inner(),
    };
    if !force {
        if let Some(prev) = &guard.signature {
            if *prev == sig {
                return (guard.overrides.clone(), guard.required_by.clone());
            }
        }
    }
    let (overrides, required_by) = build_declarations();
    guard.overrides = overrides.clone();
    guard.required_by = required_by.clone();
    guard.signature = Some(sig);
    (overrides, required_by)
}

/// Drop the cached scan so the next call rebuilds.
pub fn invalidate_cache() {
    let mut guard = match CACHE.lock() {
        Ok(g) => g,
        Err(p) => p.into_inner(),
    };
    guard.signature = None;
    guard.overrides.clear();
    guard.required_by.clear();
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::memory::common::TEST_ENV_LOCK;
    use std::sync::MutexGuard;
    use tempfile::TempDir;

    struct WyldeRootEnv {
        _guard: MutexGuard<'static, ()>,
        _td: TempDir,
        prior_root: Option<std::ffi::OsString>,
        prior_services: Option<std::ffi::OsString>,
    }

    impl WyldeRootEnv {
        fn new() -> Self {
            let guard = TEST_ENV_LOCK.lock().unwrap_or_else(|p| p.into_inner());
            let td = TempDir::new().expect("tempdir");
            let prior_root = std::env::var_os("WYLDE_ROOT");
            // The `Services/` bucket walk honours a `WYLDE_SERVICES` override
            // (`wylde_stack::roster::resolve_bucket_dir`), which a real dev
            // machine sets to its live install. Clear it so the scan resolves
            // under the tempdir; without this the discovery would read the
            // developer's actual `Services/` and the test wouldn't be hermetic.
            let prior_services = std::env::var_os("WYLDE_SERVICES");
            std::env::set_var("WYLDE_ROOT", td.path());
            std::env::remove_var("WYLDE_SERVICES");
            invalidate_cache();
            Self {
                _guard: guard,
                _td: td,
                prior_root,
                prior_services,
            }
        }

        fn root(&self) -> &Path {
            self._td.path()
        }
    }

    impl Drop for WyldeRootEnv {
        fn drop(&mut self) {
            invalidate_cache();
            match self.prior_root.take() {
                Some(v) => std::env::set_var("WYLDE_ROOT", v),
                None => std::env::remove_var("WYLDE_ROOT"),
            }
            match self.prior_services.take() {
                Some(v) => std::env::set_var("WYLDE_SERVICES", v),
                None => std::env::remove_var("WYLDE_SERVICES"),
            }
        }
    }

    fn write_manifest(at: &Path, service: &str, models: &[(&str, &str)]) {
        std::fs::create_dir_all(at.parent().unwrap()).unwrap();
        let arr: Vec<_> = models
            .iter()
            .map(|(id, kind)| serde_json::json!({"id": id, "kind": kind, "required": true}))
            .collect();
        let m = serde_json::json!({
            "name": service,
            "models": arr,
        });
        std::fs::write(at, serde_json::to_string_pretty(&m).unwrap()).unwrap();
    }

    #[test]
    fn empty_root_returns_empty_decls() {
        let _env = WyldeRootEnv::new();
        let (o, r) = load_declarations(true);
        assert!(o.is_empty());
        assert!(r.is_empty());
    }

    #[test]
    fn services_bucket_manifest_is_seen_by_the_model_registry() {
        // #125B falsification: a service dropped into the `Services/` bucket
        // that declares a model must be visible to the model registry — with no
        // code edit. Before deriving the roots from `discovered_folders` the
        // scan walked a hand-kept list of pre-cutover names that did NOT include
        // `Services/`, so this manifest was silently invisible; reverting to
        // that literal turns this red. Mirrors the discovery seam
        // `wylde_updater/tests/whole_stack_coverage.rs` drives.
        let env = WyldeRootEnv::new();
        write_manifest(
            &env.root().join("Services/wylde-synthetic/manifest.json"),
            "wylde-synthetic",
            &[("acme/embed-v1", "embed")],
        );
        let (overrides, required) = load_declarations(true);
        assert_eq!(
            overrides.get("acme/embed-v1"),
            Some(&Kind::Embed),
            "a Services/ manifest's model must reach the registry",
        );
        assert_eq!(
            required.get("acme/embed-v1"),
            Some(&vec!["wylde-synthetic".to_owned()])
        );
    }

    #[test]
    fn manifest_at_top_level_of_service_is_picked_up() {
        let env = WyldeRootEnv::new();
        write_manifest(
            &env.root().join("Services/Voice/manifest.json"),
            "voice",
            &[
                ("openai/whisper-small", "stt"),
                ("rhasspy/piper-voices", "tts"),
            ],
        );
        let (overrides, required) = load_declarations(true);
        assert_eq!(overrides.get("openai/whisper-small"), Some(&Kind::Stt));
        assert_eq!(overrides.get("rhasspy/piper-voices"), Some(&Kind::Tts));
        assert_eq!(
            required.get("openai/whisper-small"),
            Some(&vec!["voice".to_owned()])
        );
    }

    #[test]
    fn underscore_prefixed_bucket_children_are_skipped() {
        let env = WyldeRootEnv::new();
        // `discovered_folders` excludes `_`/`.`-prefixed immediate children of
        // the bucket (private/dotfiles) — so a `Services/_staging/` never
        // reaches the model registry, mirroring the daemon's own discovery.
        write_manifest(
            &env.root().join("Services/_staging/manifest.json"),
            "staging",
            &[("ignored/model", "llm")],
        );
        let (overrides, _) = load_declarations(true);
        assert!(overrides.is_empty());
    }

    #[test]
    fn bad_kind_dropped_silently() {
        let env = WyldeRootEnv::new();
        write_manifest(
            &env.root().join("Services/Voice/manifest.json"),
            "voice",
            &[("openai/whisper-small", "audio")], // unknown kind
        );
        let (overrides, _) = load_declarations(true);
        assert!(overrides.is_empty());
    }

    #[test]
    fn first_declaration_wins_on_conflict() {
        let env = WyldeRootEnv::new();
        write_manifest(
            &env.root().join("Services/svc-a/manifest.json"),
            "svc-a",
            &[("multi/model", "stt")],
        );
        write_manifest(
            &env.root().join("Services/svc-b/manifest.json"),
            "svc-b",
            &[("multi/model", "llm")],
        );
        let (overrides, required) = load_declarations(true);
        // `discovered_folders` returns the bucket's children sorted by path, so
        // `svc-a` scans before `svc-b`; the first declaration's kind wins.
        assert_eq!(overrides.get("multi/model"), Some(&Kind::Stt));
        // Both services are still recorded in required_by.
        let req = required.get("multi/model").cloned().unwrap_or_default();
        assert_eq!(req.len(), 2);
        assert!(req.contains(&"svc-a".to_owned()));
        assert!(req.contains(&"svc-b".to_owned()));
    }

    #[test]
    fn manifest_with_no_models_array_is_tolerated() {
        let env = WyldeRootEnv::new();
        let path = env.root().join("Services/Voice/manifest.json");
        std::fs::create_dir_all(path.parent().unwrap()).unwrap();
        std::fs::write(&path, r#"{"name": "voice"}"#).unwrap();
        let (overrides, _) = load_declarations(true);
        assert!(overrides.is_empty());
    }

    #[test]
    fn bad_json_is_tolerated() {
        let env = WyldeRootEnv::new();
        let path = env.root().join("Services/Voice/manifest.json");
        std::fs::create_dir_all(path.parent().unwrap()).unwrap();
        std::fs::write(&path, "not json").unwrap();
        let (overrides, _) = load_declarations(true);
        assert!(overrides.is_empty());
    }

    #[test]
    fn cache_invalidate_forces_rebuild() {
        let env = WyldeRootEnv::new();
        write_manifest(
            &env.root().join("Services/Voice/manifest.json"),
            "voice",
            &[("openai/whisper-small", "stt")],
        );
        let _ = load_declarations(true);
        invalidate_cache();
        write_manifest(
            &env.root().join("Services/Voice/manifest.json"),
            "voice",
            &[
                ("openai/whisper-small", "stt"),
                ("rhasspy/piper-voices", "tts"),
            ],
        );
        let (overrides, _) = load_declarations(false);
        // After invalidation + manifest change, both models should appear.
        assert_eq!(overrides.len(), 2);
    }
}

File: Core/harness/model_registry/_service_manifests.py
"""model_registry/_service_manifests.py — read ``models`` declarations from services.

Each top-level service folder under the Wylde root may ship a
``manifest.json`` with a ``models`` array, e.g.::

    {
      "name": "VoiceAssistant",  # wylde-check: dead-ref-ok
      "models": [
        {"id": "openai/whisper-small", "kind": "stt", "required": true},
        {"id": "rhasspy/piper-voices", "kind": "tts", "required": true}
      ]
    }

This module collects every such declaration and returns:

* an ``overrides`` dict (model id → kind), used by the HF scanner to win
  over the heuristic, and
* a ``required_by`` dict (model id → list of service names) so the GUI
  can show which subsystem owns each model.

Manifests without a ``models`` key are silently skipped — most services
predate this contract and need no migration. The scanner also walks
``Core/harness/tooling/tools/<group>/<id>/manifest.json`` but those are
*tool* manifests, not *service* manifests; they're skipped by virtue of
the search root.
"""

from __future__ import annotations

import json
import logging
import os
import threading
from pathlib import Path
from typing import Dict, List, Optional, Set, Tuple

from ._types import KIND_VALUES, Kind

logger = logging.getLogger("wylde.harness.model_registry.service_manifests")

# model_registry/_service_manifests.py → model_registry → harness → Core → Wylde
_WYLDE_ROOT = Path(__file__).resolve().parents[3]

# Top-level dirs to scan. Skipping ``Core`` (no service manifests there) and
# ``_legacy`` (per the Wylde user's note: stay out of _legacy/). Anything new added
# to the Wylde root that ships a service manifest needs to be added here.
_SERVICE_ROOTS: Tuple[str, ...] = (
    "Voice",
    "VoiceAssistant",  # wylde-check: dead-ref-ok
    "Extensions",
    "Gateway",
    "device_gate",
    "N8N",
)

_lock = threading.Lock()
_cached_signature: Optional[Tuple[Tuple[str, float, int], ...]] = None
_cached_overrides: Dict[str, Kind] = {}
_cached_required_by: Dict[str, List[str]] = {}


def _candidate_manifests() -> List[Path]:
    """Find every service ``manifest.json`` under the recognised roots.

    A service manifest sits at ``<service>/manifest.json`` (top-level of the
    service folder) or at ``<service>/<sub>/manifest.json`` for nested
    services like ``Voice/_wylde_voice``. We allow both; tool manifests
    under ``Core/harness/tooling/tools/`` are not reached because we don't
    scan ``Core``.
    """
    paths: List[Path] = []
    seen: Set[Path] = set()
    for root_name in _SERVICE_ROOTS:
        root = _WYLDE_ROOT / root_name
        if not root.is_dir():
            continue
        # Direct child manifest first.
        direct = root / "manifest.json"
        if direct.is_file() and direct not in seen:
            paths.append(direct)
            seen.add(direct)
        # One level deep for the ``Voice/_wylde_voice``-style services.
        try:
            for child in root.iterdir():
                if not child.is_dir():
                    continue
                if child.name.startswith("__") or child.name.startswith("."):
                    continue
                m = child / "manifest.json"
                if m.is_file() and m not in seen:
                    paths.append(m)
                    seen.add(m)
        except OSError:
            continue
    return paths


def _scan_signature(paths: List[Path]) -> Tuple[Tuple[str, float, int], ...]:
    sigs: List[Tuple[str, float, int]] = []
    for p in paths:
        try:
            st = p.stat()
        except OSError:
            continue
        sigs.append((str(p), st.st_mtime, st.st_size))
    sigs.sort()
    return tuple(sigs)


def _coerce_kind(value: object) -> Optional[Kind]:
    if isinstance(value, str) and value in KIND_VALUES:
        return value  # type: ignore[return-value]
    return None


def _read_one(manifest: Path) -> Tuple[str, List[Tuple[str, Kind]]]:
    """Return (service_name, [(model_id, kind), ...]) for one manifest.

    Service name comes from the manifest's ``name`` field, falling back to
    the parent folder. Bad JSON or missing ``models`` arrays are tolerated
    and produce an empty model list, never an exception.
    """
    try:
        data = json.loads(manifest.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError) as exc:
        logger.debug("service_manifests: skip %s (%s)", manifest, exc)
        return manifest.parent.name, []
    if not isinstance(data, dict):
        return manifest.parent.name, []
    service = str(data.get("name") or manifest.parent.name)
    models_raw = data.get("models")
    if not isinstance(models_raw, list):
        return service, []
    out: List[Tuple[str, Kind]] = []
    for spec in models_raw:
        if not isinstance(spec, dict):
            continue
        model_id = str(spec.get("id") or "").strip()
        kind = _coerce_kind(spec.get("kind"))
        if not model_id or kind is None:
            logger.debug(
                "service_manifests: %s: dropped entry %r (need id and recognised kind)",
                manifest,
                spec,
            )
            continue
        out.append((model_id, kind))
    return service, out


def _build() -> Tuple[Dict[str, Kind], Dict[str, List[str]]]:
    overrides: Dict[str, Kind] = {}
    required_by: Dict[str, List[str]] = {}
    for manifest in _candidate_manifests():
        service, decls = _read_one(manifest)
        for model_id, kind in decls:
            existing = overrides.get(model_id)
            if existing is not None and existing != kind:
                logger.warning(
                    "service_manifests: %s declares %s as %s but a prior service "
                    "declared %s; keeping first.",
                    service,
                    model_id,
                    kind,
                    existing,
                )
            else:
                overrides[model_id] = kind
            required_by.setdefault(model_id, []).append(service)
    return overrides, required_by


def load_declarations(
    *, force: bool = False
) -> Tuple[Dict[str, Kind], Dict[str, List[str]]]:
    """Return cached ``(overrides, required_by)`` from all service manifests.

    Cached on the manifests' (path, mtime, size) signature, same template
    as the HF scanner and ``tool_registry``. ``force=True`` rebuilds.
    """
    global _cached_signature, _cached_overrides, _cached_required_by
    paths = _candidate_manifests()
    sig = _scan_signature(paths)
    with _lock:
        if not force and _cached_signature is not None and sig == _cached_signature:
            return dict(_cached_overrides), {
                k: list(v) for k, v in _cached_required_by.items()
            }
        overrides, required_by = _build()
        _cached_signature = sig
        _cached_overrides = overrides
        _cached_required_by = required_by
        return dict(overrides), {k: list(v) for k, v in required_by.items()}


def invalidate_cache() -> None:
    global _cached_signature, _cached_overrides, _cached_required_by
    with _lock:
        _cached_signature = None
        _cached_overrides = {}
        _cached_required_by = {}


def wylde_root() -> Path:
    """For diagnostics / tests — show what path we resolve as the repo root."""
    override = os.getenv("WYLDE_ROOT")
    if override:
        return Path(override).expanduser()
    return _WYLDE_ROOT


__all__ = ["load_declarations", "invalidate_cache", "wylde_root"]
